### PR TITLE
DrivAerML: LR warmup on EMA+gc=0.5 champion (5-ep & 10-ep)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

LR warmup combined with the full EMA+gc champion platform has never been tested on DrivAerML. Two prior warmup-only runs diverged without EMA: franky (#3148, 10-ep warmup + gc=1.0, reached 11.2% then diverged) and chopper (#3137, 5-ep warmup + gc=1.0). The hypothesis is that EMA's exponential moving average smooths the warmup→full-LR transition, allowing the model to survive where warmup alone failed.

EMA (decay=0.9995) + gradient clipping (gc=0.5) is the proven stability package for DrivAerML — it took eren from divergence at ep28 to a new best of 3.833%. Adding warmup on top of this stable platform may let the model settle into a better initial basin before the cosine schedule kicks in, potentially improving convergence below 3.833%.

Key question: Does warmup+EMA survive and improve where warmup alone diverged?

## Instructions

Run two trials back-to-back (or in parallel if two GPUs are available). Both use the full EMA+gc=0.5 champion config, differing only in warmup length.

**Trial 1 — 5-epoch warmup:**
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 5e-4 \
  --cosine-t-max 30 \
  --grad-clip 0.5 \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --ema-decay 0.9995 \
  --warmup-epochs 5 \
  --wandb_group dm-warmup-ema-gc
```

**Trial 2 — 10-epoch warmup:**
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 5e-4 \
  --cosine-t-max 30 \
  --grad-clip 0.5 \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --ema-decay 0.9995 \
  --warmup-epochs 10 \
  --wandb_group dm-warmup-ema-gc
```

Report the best `val_primary/surface_rel_l2_pct` for each trial, the W&B run IDs, and whether either run diverged. Note the epoch at which the best val metric was achieved.

## Baseline

Current best DrivAerML metrics (PR #3072, eren — EMA=0.9995 + gc=0.5, 4L/512d/8H, AdamW lr=5e-4, T_max=30):
- **val_primary/surface_rel_l2_pct: 3.833%** (epoch 511)
- test surface_rel_l2_pct: 4.685%
- W&B run: `ncl1dh88` (eren/dm-ema-9995-gc05)
- External target: <3.71% (AB-UPT)

Reproduce baseline:
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 5e-4 \
  --cosine-t-max 30 \
  --grad-clip 0.5 \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --ema-decay 0.9995
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT sanji:

## Results

### Trial 1 — 5-epoch warmup + EMA=0.9995 + gc=0.5

| Metric | Baseline | 5-ep warmup | Delta |
|--------|----------|-------------|-------|
| val_primary/surface_rel_l2_pct | 3.833% (ep 511) | 11.325% (ep 31) | +7.49% (worse) |
| test_primary/surface_rel_l2_pct | 4.685% | 11.804% | +7.12% (worse) |

- **W&B run:** ulog3zq2
- **Peak memory:** ~6.0 GB
- **Command:** `cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --grad-clip 0.5 --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --ema-decay 0.9995 --warmup-epochs 5`
- **Outcome:** Diverged. Model improved steadily during warmup (epochs 1-5), continued improving to val=11.33% at epoch 31, then collapsed. By epoch 50, grad norms were 2.4x normal with 97% of steps clipped. By epoch 100, grad norms=Infinity, val=66.33%. The 5-epoch warmup was too short — the model transitioned too quickly to full LR, and even EMA+gc couldn't prevent instability.

### Trial 2 — 10-epoch warmup + EMA=0.9995 + gc=0.5

| Metric | Baseline | 10-ep warmup | Delta |
|--------|----------|--------------|-------|
| val_primary/surface_rel_l2_pct | 3.833% (ep 511) | 3.918% (ep 496) | +0.09% (slightly worse) |
| test_primary/surface_rel_l2_pct | 4.685% | 4.818% | +0.13% (slightly worse) |

- **W&B run:** 5ltdja3j
- **Peak memory:** ~6.0 GB
- **Command:** `cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --grad-clip 0.5 --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --ema-decay 0.9995 --warmup-epochs 10`
- **Outcome:** Survived and converged well. The 10-epoch linear warmup (LR from 5e-6 to 5e-4) gave the model enough settling time before the cosine schedule kicked in. Convergence was steady and stable throughout — grad norms dropped from 0.58 at ep50 to 0.087 at ep450, gc clipping fell from 49% to <1%. Reached 3.918% at epoch 496, just 0.09% short of the 3.833% baseline. The training was limited by timeout — still improving at termination.

### What happened

**10-ep warmup survived; 5-ep warmup diverged.** The key finding is that EMA+gc can stabilize the warmup→full-LR transition, but only with a long enough warmup. 5 epochs was insufficient — the model transitioned to full 5e-4 LR before EMA weights (still warming up via the timm adaptive formula) could provide adequate smoothing. 10 epochs gave both the LR schedule and EMA decay enough time to reach equilibrium.

The 10-ep warmup result (3.918% at ep496) came very close to the no-warmup baseline (3.833% at ep511) but didn't beat it. The convergence trajectory was slightly slower — the warmup essentially "used up" 10 epochs of training budget on sub-optimal LR. However, the run was still improving when timeout hit, so with a longer budget it might have matched or beaten the baseline.

The answer to the PR's key question: **Yes, warmup+EMA survives where warmup alone diverged (franky #3148, chopper #3137).** But it doesn't improve on the EMA+gc champion — warmup is neutral-to-slightly-negative when the stability package is already in place.

### Suggested follow-ups

1. **Longer training budget for 10-ep warmup:** The run was still improving at termination. A 1500-epoch budget might close the 0.09% gap.
2. **Warmup sweep (7, 8, 15, 20 epochs):** 5-ep diverged, 10-ep survived. There's likely a critical warmup length between 5-10 where stability transitions. Understanding this boundary could inform other schedule experiments.
3. **Gentler warmup start factor:** The current start_factor=0.01 starts at LR=5e-6. A warmer start (start_factor=0.1, LR=5e-5) might let 5-ep warmup survive while a colder start (start_factor=0.001) might help even shorter warmups.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3171 because ADVISOR: LR warmup dead for DM champion config. 5-ep warmup: 11.325% ep31 then collapsed (W&B: ulog3zq2) — warmup too short, abrupt LR ramp destabilizes. 10-ep warmup: 3.918% ep496, plateaued (W&B: 5ltdja3j) — stable but didn't beat 3.833%, convergence analysis shows no further descent potential (last 50 epochs: mean 4.02%, no trend). Warmup is neutral-to-harmful when EMA+gc=0.5 already provides stability. Adding to dead ends: warmup with EMA+gc..
